### PR TITLE
add-shopping-cart

### PR DESCRIPTION
**Pull request by AndreyPolezhaiev**

### CODE REVIEW

File: src/main/java/com/polezhaiev/shop/model/CartItem.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+@Entity(name = "carts_items")
+@Data
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @EqualsAndHashCode.Exclude
+    @ToString.Exclude
+    @OneToOne
```
Comment (Rommelua) [inline src/main/java/com/polezhaiev/shop/model/CartItem.java]:
```suggestion
    @ManyToOne
```

File: src/main/java/com/polezhaiev/shop/model/CartItem.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    @JoinColumn(name = "shopping_cart_id",
+            nullable = false,
+            columnDefinition = "int default 0")
+    private ShoppingCart shoppingCart;
+
+    @EqualsAndHashCode.Exclude
+    @ToString.Exclude
+    @OneToOne
```
Comment (Rommelua) [inline src/main/java/com/polezhaiev/shop/model/CartItem.java]: ManyToOne

File: src/main/java/com/polezhaiev/shop/service/shopcart/ShoppingCartService.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package com.polezhaiev.shop.service.shopcart;
+
+import com.polezhaiev.shop.dto.shopcart.CartItemRequestDto;
+import com.polezhaiev.shop.dto.shopcart.CartItemUpdateRequestDto;
+import com.polezhaiev.shop.dto.shopcart.ShoppingCartResponseDto;
+
+public interface ShoppingCartService {
+    ShoppingCartResponseDto getShoppingCart(Long id);
+
+    ShoppingCartResponseDto addBookToShopCart(Long id, CartItemRequestDto requestDto);
```
Comment (Rommelua) [inline src/main/java/com/polezhaiev/shop/service/shopcart/ShoppingCartService.java]:
```suggestion
    ShoppingCartResponseDto addBookToShoppingCart(Long id, CartItemRequestDto requestDto);
```

File: src/main/java/com/polezhaiev/shop/service/shopcart/ShoppingCartService.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import com.polezhaiev.shop.dto.shopcart.CartItemRequestDto;
+import com.polezhaiev.shop.dto.shopcart.CartItemUpdateRequestDto;
+import com.polezhaiev.shop.dto.shopcart.ShoppingCartResponseDto;
+
+public interface ShoppingCartService {
+    ShoppingCartResponseDto getShoppingCart(Long id);
+
+    ShoppingCartResponseDto addBookToShopCart(Long id, CartItemRequestDto requestDto);
+
+    ShoppingCartResponseDto updateQuantityOfABook(Long id, CartItemUpdateRequestDto requestDto);
```
Comment (Rommelua) [inline src/main/java/com/polezhaiev/shop/service/shopcart/ShoppingCartService.java]:
```suggestion
    ShoppingCartResponseDto updateCartItem(Long id, CartItemUpdateRequestDto requestDto);
```

File: src/main/java/com/polezhaiev/shop/service/shopcart/ShoppingCartService.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+    void deleteById(Long id);
```
Comment (Rommelua) [inline src/main/java/com/polezhaiev/shop/service/shopcart/ShoppingCartService.java]:
```suggestion
    void deleteShoppingCartById(Long id);

```

File: src/main/java/com/polezhaiev/shop/model/CartItem.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import jakarta.persistence.Id;
+import jakarta.persistence.JoinColumn;
+import jakarta.persistence.OneToOne;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.ToString;
+import org.hibernate.annotations.SQLDelete;
+
+@SQLDelete(sql = "UPDATE carts_items SET is_deleted = true WHERE id=?")
+@Entity(name = "carts_items")
```
Comment (Rommelua) [inline src/main/java/com/polezhaiev/shop/model/CartItem.java]:
```suggestion
@SQLDelete(sql = "UPDATE carts_items SET is_deleted = true WHERE id=?")
@Where(clause = "is_deleted = false")
@Entity(name = "carts_items")
```

File: src/main/java/com/polezhaiev/shop/service/shopcart/impl/ShoppingCartServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    private final CartItemRepository cartItemRepository;
+
+    @Override
+    public ShoppingCartResponseDto getShoppingCart(Long id) {
+        ShoppingCart shoppingCart = shoppingCartRepository.getById(id);
+        Set<CartItem> notDeletedItems = shoppingCart.getCartItems()
+                .stream()
+                .filter(i -> !i.isDeleted())
+                .collect(Collectors.toSet());
+        shoppingCart.setCartItems(notDeletedItems);
```
Comment (Rommelua) [inline src/main/java/com/polezhaiev/shop/service/shopcart/impl/ShoppingCartServiceImpl.java]:
```suggestion

```

File: src/main/java/com/polezhaiev/shop/service/shopcart/impl/ShoppingCartServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartResponseDto addBookToShopCart(Long id, CartItemRequestDto requestDto) {
+        ShoppingCart shoppingCart = shoppingCartRepository.getById(id);
+        CartItem cartItem = shoppingCartMapper.toModel(requestDto);
```
Comment (Rommelua) [inline src/main/java/com/polezhaiev/shop/service/shopcart/impl/ShoppingCartServiceImpl.java]:
```suggestion
        CartItem cartItem = cartItemMapper.toModel(requestDto);
```

**Comment by AndreyPolezhaiev** [comment]
Thank you. I hadn't  known why I can't show a shopping cart with deleted cart items, and I needed the "Where()" condition, but I forgot

**Review by koval-kn — APPROVED** [review]

**Pull request merged**